docs/proposals/thin-volumes.md

```diff
@@ -0,0 +1,189 @@
+    string device_class = 4;
++   string thin_pool = 5;
+}
+```
+- For VGService
+``` go=
+message GetLVListRequest {
+    string device_class = 1;
++   string thin_pool = 2;
+}
+
+message GetFreeBytesRequest {
+    string device_class = 1;
++   string thin_pool = 2;
+}
+
+message WatchItem {
+    uint64 free_bytes = 1;  // Free space of the volume group in bytes.
+    string device_class = 2;
+    uint64 size_bytes = 3;  // Size of the volume group in bytes.
++   string thin_pool = 4;
++   uint64 overprovision_bytes = 5; // Size up and above thin pool capacity
+}
+```
+- Resize and remove of Logical Volumes will use the same API as the underlying calls on volume group or thinpool only needs Logical Volume ID.
+
+
+#### Monitoring thin-pools
+- We need to constantly monitor the thin-pools and ensure that it does not run out of space.
+- `/etc/lvm/lvm.conf` settings can allow auto growth of the thin pool when required. By default, the threshold is 100% which means that the pool will not grow.
+- If we set this to, 75%, the Thin Pool will autoextend when the pool is 75% full. It will increase by the default percentage of 20% if the value is not changed. We can see these settings using the command grep against the file.
+
+```go=
+# grep -E ‘^\s*thin_pool_auto’ /etc/lvm/lvm.conf
+ thin_pool_autoextend_threshold = 100
+ thin_pool_autoextend_percent = 20
+```
+- We assume that user has the threshold of 100% with no autogrow.
+- Topolvm will regularly monitor the used thin-pool space for `pool data` and `metadata`.
+- These usages will be exported as metrics so that admin/opeartor can analyse them and take precautionary steps whens usage exeeds certain threshold limit.
+- Command to be used for montoring:
+``` go=
+   lvs
+   LV          VG   Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
+   lv1         vg1  -wi-ao---- 600.00m
+   thin_volume vg1  Vwi-a-tz--   1.00g tpool        3.20
+   tpool       vg1  twi-aotz-- 200.00m              16.38  1.37
+```
+- Topolvm will only monitor the thin-pools provided in the lvmd.yaml config file.
+
+#### Capacity awareness
+- Topolvm uses capacity awareness to ensure that new pod is created on the node where available capacity of the volume group is larger than the requested capacity.`capacity.topolvm.cybozu.com/<device-class>` annotation is added on each node which consists of the available capacity on the volume group.
+- Creation of thin logical volumes doesn't update the available device class capacity until the user actually uses storage on the thin volumes.
+- So we need to ensure that all the thin logical volumes don't get created on the same node.
+- This can be done by using the `overprovisionRatio`.
+- Value of `capacity.topolvm.cybozu.com/<device-class>_mlog<thin-pool>` annotation will be calculated as:
+  `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-sizes-in-thin-pool)`
+- Available thin pools on the volume groups will be obtained from lvmd.yaml config file.
```

Comment:
Let's consider below example for a single node to read in conjunction with above responses:

## Setup:
1. Sample lvmd config with thin-pool in an existing volume group `vg0`
``` yaml
device-classes:
  - name: ssd
    volume-group: vg0
    default: true
    overprovision-ratio: 10.0
    thin-pools:
        -name: pool0
```
2. Total size of `vg0` is 50G and 10G is already used for thick volumes
3. We (~admin) created `vg0/pool0` of size 10G and with overprovision-ratio of `10.0` we respect thin-lv creations amounting upto 10.0 X 10G = 100G virtual size
- Now the annotations on this node will be as follows:
```
capacity.topolvm.cybozu.com/ssd: 30G
capacity.topolvm.cybozu.com/ssd_mlogpool0: 100G
```
30G --> 50G (vg_size) - 10G (thick-lvs) - 10G (thin-pool)
100G --> 100G virtual size as mentioned above since there's no thin-lv created on this pool yet

## Request:

a. thick lv creation:

1. below params exists on storage class (sc1) for creation of thick volumes from a specific device class
```
topolvm.cybozu.com/device-class: ssd
```
2. pvc request has come up on this sc1 for 5G and topolvm created it and updated metrics & annotations
```
capacity.topolvm.cybozu.com/ssd: 25G
capacity.topolvm.cybozu.com/ssd_mlogpool0: 100G (same no change)
```

b. thin lv creation:
1. below params exists on storage class (sc2) for creation of thin volumes from a specific device class & pool
```
topolvm.cybozu.com/device-class: ssd
topolvm.cybozu.com/thin-pool-name: pool0
```
2. pvc request has come up on this sc2 for 50G and topolvm created it and updated metrics & annotations
```
capacity.topolvm.cybozu.com/ssd: 25G (same no change)
capacity.topolvm.cybozu.com/ssd_mlogpool0: 50G
```
50G --> 100G (size with overprovision) - 50 G (which is just created)
3. another pvc request on sc2 for 30G and below will be the updated annotations
```
capacity.topolvm.cybozu.com/ssd: 25G (same no change)
capacity.topolvm.cybozu.com/ssd_mlogpool0: 20G
```
20G --> 100G (size with overprovision) - 50 G (existing) - 30G (current)

Reply:
@leelavg 
Thank you for sharing your idea. I understand. I reply after discussion Open Questions.